### PR TITLE
[SMTChecker] Add CHC constructor/interface/error blocks

### PR DESCRIPTION
**Pull request by leonardoalt**
Depends on #7062

### CODE REVIEW

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -154,6 +215,71 @@ bool CHC::shouldVisit(FunctionDefinition const& _function) const
...
+{
+	auto boolSort = make_shared<smt::Sort>(smt::Kind::Bool);
+	auto const& funType = dynamic_cast<FunctionType const&>(*_function.type());
+	return make_shared<smt::FunctionSort>(
+		smt::smtSort(funType.parameterTypes()),
+		boolSort
+	);
+}
+
+shared_ptr<smt::SymbolicFunctionVariable> CHC::createBlock(smt::SortPointer _sort, string _name)
```
Comment (erak) [inline libsolidity/formal/CHC.cpp]:
```suggestion
shared_ptr<smt::SymbolicFunctionVariable> CHC::createBlock(smt::SortPointer _sort, string const& _name)
```

File: libsolidity/formal/CHC.cpp [hunk 2]
```diff
@@ -61,11 +61,62 @@ bool CHC::visit(ContractDefinition const& _contract)
...
 		return false;
 
+	m_contract = &_contract;
+
 	reset();
 
 	if (!SMTEncoder::visit(_contract))
 		return false;
 
+	for (auto const& contract: _contract.annotation().linearizedBaseContracts)
```
Comment (erak) [inline libsolidity/formal/CHC.cpp]: Maybe pull this block into a seperate function `readStateVariables` ?

File: libsolidity/formal/CHC.cpp [hunk 3]
```diff
@@ -61,11 +61,62 @@ bool CHC::visit(ContractDefinition const& _contract)
...
+		for (auto var: contract->stateVariables())
+			if (*contract == _contract || var->isVisibleInDerivedContracts())
+				m_stateVariables.push_back(var);
+
+	for (auto const& var: m_stateVariables)
```
Comment (erak) [inline libsolidity/formal/CHC.cpp:70]: Same here: `updateStateSorts` ? Or is this function going to be refactored anyways?

**Comment by erak** [comment]
@leonardoalt I would tend to approve this PR, but I'm wondering if the code in`visit(ContractDefinition...)` should be pulled into functions?

File: libsolidity/formal/CHC.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp]: Makes sense

File: libsolidity/formal/CHC.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:70]: Here I'm not sure, since it's not used anywhere else

**Comment by leonardoalt** [comment]
I think retrieving the state variables could have its own function since SMTEncoder also does that, but the other code runs only in that function, so I personally don't see the need for another function. I can do it anyway if required.

File: libsolidity/formal/CHC.cpp (hunk 3 above)
Comment (erak) [reply libsolidity/formal/CHC.cpp:70]: ok!

**Comment by erak** [comment]
Ok, then just extract that one function.

**Comment by leonardoalt** [comment]
@erak updated

File: libsolidity/formal/SMTEncoder.cpp [hunk 4]
```diff
@@ -1321,6 +1319,16 @@ Expression const* SMTEncoder::leftmostBase(IndexAccess const& _indexAccess)
 	return base;
 }
 
+vector<VariableDeclaration const*> SMTEncoder::stateVariables(ContractDefinition const& _contract)
```
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp]: This is the new function @erak

Comment (chriseth) [reply libsolidity/formal/SMTEncoder.cpp]: Shouldn't this go into AST.h/cpp?

File: libsolidity/formal/CHC.cpp [hunk 5]
```diff
@@ -61,11 +61,59 @@ bool CHC::visit(ContractDefinition const& _contract)
...
+	// TODO create static instances for Bool/Int sorts in SolverInterface.
+	auto boolSort = make_shared<smt::Sort>(smt::Kind::Bool);
+	auto errorFunctionSort = make_shared<smt::FunctionSort>(
+		vector<smt::SortPointer>(),
+		boolSort
+	);
+	m_errorPredicate = createBlock(errorFunctionSort, "error");
+
+	// If the contract has a constructor it is handled as a function.
+	// Otherwise we zero-initialize all state vars.
```
Comment (chriseth) [inline libsolidity/formal/CHC.cpp:89]: Don't we need to do that in both cases?

File: libsolidity/formal/SMTEncoder.cpp (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Makes sense, updated

File: libsolidity/formal/CHC.cpp (hunk 5 above)
Comment (leonardoalt) [reply libsolidity/formal/CHC.cpp:89]: Not sure, BMC wouldn't really make use of initialized state vars while checking functions

**Comment by leonardoalt** [comment]
https://github.com/ethereum/solidity/pull/7170 should be merged first to get the `stateVariablesWithInheritance` stuff

**Comment by ekpyron** [comment]
Needs rebase.

**Review by ekpyron — APPROVED** [review]
Since @leonardoalt assured me that everything I found will be fixed/changed in the subsequent PRs, I'm fine with merging now :-).

**Comment by leonardoalt** [comment]
Those were
- Zero initializing state variables also before the explicit constructor (still unimplemented for both BMC and CHC)
- `functionSort` should also have return parameters. The next PR that actually implements functions does that.

**Pull request merged**